lib/Conversion/CalyxNative/CalyxNative.cpp

```diff
@@ -0,0 +1,150 @@
+
+#include "../PassDetail.h"
+
+#include "circt/Conversion/CalyxNative.h"
+#include "circt/Dialect/Calyx/CalyxEmitter.h"
+#include "circt/Dialect/SV/SVDialect.h"
+#include "circt/Dialect/Seq/SeqOps.h"
+#include "mlir/Parser/Parser.h"
+#include "mlir/Support/FileUtilities.h"
+#include "llvm/Support/MemoryBuffer.h"
+#include "llvm/Support/Path.h"
+#include "llvm/Support/Process.h"
+#include "llvm/Support/ToolOutputFile.h"
+
+using namespace mlir;
+using namespace circt;
+
+/// ConversionPatterns.
+
+/// Pass entrypoint.
+
+namespace {
+class CalyxNativePass : public CalyxNativeBase<CalyxNativePass> {
+public:
+  void runOnOperation() override;
+};
+} // end anonymous namespace
+
+void CalyxNativePass::runOnOperation() {
+  ModuleOp root = getOperation();
+
+  // User must specify location of the calyx primitive library.
+  if (primitiveLib == "") {
+    // XXX(rachitnigam): Probably a bad idea to hard code the name of the flag.
+    // Is there a better way?
+    root.emitError("primitive library not specified. Please specify it using "
+                   "--lower-using-calyx-native=\"primitives=<path>\"");
+    return;
+  }
+
+  SmallString<32> execName = llvm::sys::path::filename("calyx");
+  auto exeMb = llvm::sys::findProgramByName(execName);
+  // If cannot find the executable, then nothing to do, return.
+  if (!exeMb) {
+    root.emitError("cannot find executable: '" + execName + "'");
+    return;
+  }
+  StringRef generatorExe = exeMb.get();
+
+  SmallVector<std::string> generatorArgs;
+  // First argument should be the executable name.
+  generatorArgs.push_back(generatorExe.str());
+
+  SmallVector<StringRef> generatorArgStrRef;
+  for (const std::string &a : generatorArgs)
+    generatorArgStrRef.push_back(a);
+
+  std::string errMsg;
+  SmallString<32> nativeInputFileName;
+  auto errCode = llvm::sys::fs::getPotentiallyUniqueTempFileName(
+      "calyxNativeTemp", StringRef(""), nativeInputFileName);
+
+  // Default error code is 0.
+  std::error_code ok;
+  if (errCode != ok) {
+    root.emitError("cannot generate a unique temporary file name");
+    return;
+  }
+
+  // Emit the current program into a file so the native compiler can operate
+  // over it.
+  auto inputFile = mlir::openOutputFile(nativeInputFileName, &errMsg);
+  if (!inputFile) {
+    root.emitError(errMsg);
+    return;
+  }
+
+  auto res = circt::calyx::exportCalyx(root, inputFile->os());
+  inputFile->os().flush();
+  if (failed(res)) {
+    return;
+  }
+
+  // Create a file for the native compiler to write the results into
+  SmallString<32> nativeOutputFileName;
+  errCode = llvm::sys::fs::getPotentiallyUniqueTempFileName(
+      "calyxNativeOutTemp", StringRef(""), nativeOutputFileName);
+  if (errCode != ok) {
+    root.emitError(
+        "cannot generate a unique temporary file name to store output");
+    return;
+  }
+
+  std::optional<StringRef> redirects[] = {
+      /*stdin=*/std::nullopt,
+      /*stdout=*/StringRef(nativeOutputFileName),
```

Comment:
While this is legal because nativeOutputFileName will outlive redirects, I am still hesitant to do it. Can we just make these std::string to avoid any lifetime bugs?

Reply:
Hm, when I do that, I get a runtime error saying `No such file found`. Can you explain a bit more about the potential lifetime problem here? I'm not a C++ expert so it would be useful to understand what could go wrong here as well.

Reply:
I was saying 

```
std::optional<std::string> redirects[] = {
  /*stdin=*/std::nullopt,
  /*stdout=*/nativeOutputFileName,
  /*stderr=*/std::nullopt,
};
```

That isn't working?

The only issue I could think of is type failures from 

`std::optional<StringRef> <-> std::optional<std::string>`

Reply:
Ah, that works! I was looking at the output of an older command. Removed a bunch of useless `StringRef` from the code now.